Implement support for the WM_Sn selection

psychon opened the pull request:
This implements https://github.com/i3/i3/issues/536. At startup, i3 will now acquire the `WM_Sn` selection. If the selection is already owned, it will exit saying that another window manager is already running. When started with `--replace`, it will instead replace the existing WM (assuming it properly follows ICCCM, but that's always the case when working with X11).

psychon committed:
Acquire the WM_Sn selection at startup

Related to https://github.com/i3/i3/issues/536

Signed-off-by: Uli Schlachter <psychon@znc.in>

psychon committed:
WM_Sn: Deal with already running WMs

When the WM_Sn selection is already owned at startup, this now either
errors out or waits for the old selection owner to exit. The "wait to
exit"-part is not yet wired up to a possible --replace command line
option.

Signed-off-by: Uli Schlachter <psychon@znc.in>

psychon committed:
Exit when we loose WM_Sn ownership

Signed-off-by: Uli Schlachter <psychon@znc.in>

psychon committed:
Add a --replace argument

Somehow I expected this to be more work and thus I left this for its own
commit. But there is not much to do...

Signed-off-by: Uli Schlachter <psychon@znc.in>

psychon committed:
Get a proper timestamp during startup

Signed-off-by: Uli Schlachter <psychon@znc.in>

Airblader commented on src/handlers.c:
```diff
@@ -1076,6 +1076,21 @@ static void handle_configure_notify(xcb_configure_notify_event_t *event) {
 
+/*
+ * Handles SelectionClear events for the root window, which are generated when
+ * we loose ownership of a selection.
```
nit: s/loose/lose/

Airblader commented on src/main.c:
```diff
@@ -656,6 +685,95 @@ int main(int argc, char *argv[]) {
+        if (selection_reply && selection_reply->owner != XCB_NONE) {
+            xcb_get_geometry_reply_t *geom_reply = NULL;
+            DLOG("waiting for old WM_Sn selection owner to exit");
+            do {
```
Do you think it'd be worth having some upper bound of how long we wait here to avoid busy-looping forever if the owner doesn't exit?

psychon replied:
Dunno. I never saw anything doing anything clever here and I never heard about any complaints that this was going wrong.

Random example of something that uses a 15 seconds timeout and sleeps for 0.1 seconds between checks: https://github.com/danakj/openbox/blob/9e8813e111cbe6c1088f6abbc771a29470f05fc2/openbox/screen.c#L132-L153

orestisfl replied:
I am in favor of the openbox approach

psychon replied:
I implemented a (completely untested) timeout of 15 seconds for the "wait for the old WM to exit" code. I hope that's what you meant.

I did not implement the "wait for a DestroyNotify event to indicate that the other WM went away"-part, since I prefer the current "use GetGeometry to check if the window went away". If you want, I can also switch from GetGeometry to waiting for a DestroyNotify event.

psychon committed:
Fix spelling

Signed-off-by: Uli Schlachter <psychon@znc.in>

psychon committed:
clang-format insists on writing it like this

Signed-off-by: Uli Schlachter <psychon@znc.in>

psychon commented:
I just found https://github.com/i3/i3/blob/next/src/main.c#L426 which should be updated in this PR.

psychon committed:
Document --replace in --help output

Signed-off-by: Uli Schlachter <psychon@znc.in>

psychon commented:
The Travis build just failed. Somehow I feel like this has nothing to do with my PR and I have no idea what is going on.

https://travis-ci.com/github/psychon/i3/jobs/490045736

```
--- command ---
13:05:26 /usr/bin/perl complete-run.pl
--- stdout ---
Running tests under Xvfb display :99
Starting 4 Xephyr instances, starting at :100...
Writing logfile to 'testsuite-2021-03-11-13-05-26-4.15-1110-g6f05a5ca/complete-run.log'...
#   Failed test 'call to scratchpad failed'
#   at /usr/src/i3/distbuild/meson-private/dist-unpack/i3-4.19.2/testcases/t/185-scratchpad.t line 492.
#          got: '1'
#     expected: '0'
Test Summary Report
-------------------
/usr/src/i3/distbuild/meson-private/dist-unpack/i3-4.19.2/testcases/t/185-scratchpad.t (Wstat: 0 Tests: 100 Failed: 1)
  Failed test:  91
Files=252, Tests=3704, 22 wallclock secs ( 0.87 usr +  0.09 sys =  0.96 CPU)
Result: FAIL
The slowest tests are:
	/usr/src/i3/distbuild/meson-private/dist-unpack/i3-4.19.2/testcases/t/301-shape.t with 6.33 seconds
	/usr/src/i3/distbuild/meson-private/dist-unpack/i3-4.19.2/testcases/t/290-keypress-numlock.t with 5.57 seconds
	/usr/src/i3/distbuild/meson-private/dist-unpack/i3-4.19.2/testcases/t/308-focus_wrapping.t with 4.07 seconds
	/usr/src/i3/distbuild/meson-private/dist-unpack/i3-4.19.2/testcases/t/258-keypress-release.t with 3.79 seconds
	/usr/src/i3/distbuild/meson-private/dist-unpack/i3-4.19.2/testcases/t/286-root-window-mouse-binding.t with 3.60 seconds
--- stderr ---
Use of uninitialized value $timingsjson in numeric gt (>) at complete-run.pl line 206.
Use of uninitialized value in subtraction (-) at /usr/share/perl/5.32/TAP/Formatter/Base.pm line 386.
Negative repeat count does nothing at /usr/share/perl/5.32/TAP/Formatter/Base.pm line 386.
-------
```

Airblader reviewed (approved):
This looks good to me, and I'm also OK with not having the timeout for now (since this is an opt-in feature I don't see this as much of a risk). Wouldn't mind if someone had a look as well, in particular I don't have capacities at the moment to test this myself. @orestisfl perhaps?

orestisfl reviewed (changes requested):
Release notes need to be updated :)

SGTM but will check over the weekend

psychon committed:
Add two lines to RELEASE-NOTES-next about this PR

One line is the more technical explanation and the other is the more
user-visible one. I do not have to tell people that these two lines go
hand-in-hand, do I? ;-)

Signed-off-by: Uli Schlachter <psychon@znc.in>

psychon commented:
I just found this in https://github.com/i3/i3/issues/3009:

> Sometimes i3 cannot change the root window’s event mask (X11 error BadAccess), implying that the old i3 process on that x server hasn’t yet been cleaned up (even though we wait for the i3 process to terminate). 

Two comments:
- Only because i3 terminates does not mean that the X11 server already handled the connection close
- This PR could help here by making the test suite use `--replace`. That should make the new i3 process wait until the old one is "really gone".

(I am writing this here so that #3009 doesn't get cluttered with comments)

psychon commented:
> SGTM but will check over the weekend

@orestisfl ping?

orestisfl reviewed (approved):
Thanks for the ping, LGTM.

I'd also like to avoid the busy-loop. 50 requests with 0.1 sleep sound good (5 seconds in total should be enough even if the other wm is in a bad state).

psychon committed:
Merge branch 'next' into WM_Sn

psychon committed:
Wait maximum of 15 seconds for old WM to exit

Signed-off-by: Uli Schlachter <psychon@znc.in>

orestisfl commented on src/main.c:
```diff
@@ -627,6 +643,22 @@ int main(int argc, char *argv[]) {
+    {
+        xcb_generic_event_t *event;
+        DLOG("waiting for PropertyNotify event\n");
+        while ((event = xcb_wait_for_event(conn)) != NULL) {
+            if (event->response_type == XCB_PROPERTY_NOTIFY) {
+                last_timestamp = ((xcb_property_notify_event_t *)event)->time;
+                free(event);
+                break;
+            }
+            free(event);
+        }
```
Is it possible to lose events that we might want to keep in this loop? Would it make sense to make this code more event-driven?

psychon replied:
Apparently so and someone already wrote similar code later in the file: https://github.com/i3/i3/blob/4c93f61353ffa905a4f15cb98524280cd92c53cb/src/main.c#L958-L975

This just discards all pending events, but handles map requests.

Since we did not select SubstructureRedirect yet, there cannot be any pending MapRequests. All other events would be discarded "down there" anyway.

orestisfl committed:
Merge branch 'next' into WM_Sn
